marklogic-data-hub/src/main/java/com/marklogic/hub/impl/HubProjectImpl.java

```diff
@@ -343,7 +357,97 @@ public void upgradeProject() throws IOException {
+                //copy jobs,staging db to hub-internal-config
+                if(path.toLowerCase().equals("job-database.json")
+                || path.toLowerCase().equals("staging-database.json")) {
+                    try {
+                        FileUtils.copyFile(f.toFile(), getHubDatabaseDir().resolve(f.getFileName()).toFile());
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+              //copy modules,final db to ml-config
+                if(path.toLowerCase().equals("modules-database.json")
+                || path.toLowerCase().equals("final-database.json")) {
+                    try {
+                        FileUtils.copyFile(f.toFile(), getUserDatabaseDir().resolve(f.getFileName()).toFile());
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+              //copy jobs,staging server to hub-internal-config
+                if(path.toLowerCase().equals("job-server.json")
+                || path.toLowerCase().equals("staging-server.json")) {
+                    try {
+                        FileUtils.copyFile(f.toFile(), getHubServersDir().resolve(f.getFileName()).toFile());
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+              //copy final server to ml-config
+                if(path.toLowerCase().equals("final-server.json")) {
+                    try {
+                        FileUtils.copyFile(f.toFile(), getUserServersDir().resolve(f.getFileName()).toFile());
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+                //move the rest of the payloads that are not part of the set to 
+                // hub-internal-config
+                if(! hubConfigFiles.contains(path.toLowerCase())) {
+                    try {
+                        FileUtils.copyInputStreamToFile(
+                                Files.newInputStream(f),
+                                getHubConfigDir().resolve(sourceDir.relativize(f)).toFile());
+                    } catch (IOException e) {
+                        throw new RuntimeException(e);
+                    }
+                }
+            });                        
+           // Files.copy(sourceDir, destDir, StandardCopyOption.REPLACE_EXISTING);
+            FileUtils.moveDirectory(sourceDir.toFile(), renamedSourceDir.toFile());
+        }       
+    }
+    
+    /*  "user-config" is the other config dir, it can contain "final-server.json", 
+     *  "staging-server.json" etc. In 3.0 ,they are deployed after hub-internal-config. But we don't
+     *  want to blindly copy all these files to ml-config as they can  cause issues during deployment.
```

Comment:
I actually think it's okay to copy all of these over - with 4.1, it's fine if the user has e.g. staging-server.json or staging-database.json in ml-config.

Reply:
Oh wait - is the issue that if the user has final-server.json in user-config and in hub-internal-config, then we'll first copy final-server.json over from hub-internal-config to ml-config, and then we'll overwrite it with the copy from user-config?

This seems like an issue with an upgrade from 3.0 to 4.0 as well - if e.g. final-database.json can be in both hub-internal-config and user-config in 3.0, but it can only be in one place in 4.0, then the two files should really be merged together, right?

Reply:
So worst case for the user, they've always got their original files in an ".old" directory right? So there's really no risk of losing anything - just some mild annoyance because some manual copying/editing may be required?

Reply:
Yeah, they can find the file in ".old" dir but I will have to check with @sbayatpur-ml if we have to merge user-config files as well for 4.1